cpp/src/binning/bin.cu

```diff
@@ -0,0 +1,215 @@
+    column_view const& input,
+    column_view const& left_edges,
+    inclusive left_inclusive,
+    column_view const& right_edges,
+    inclusive right_inclusive,
+    rmm::mr::device_memory_resource* mr)
+  {
+    // Note: We could be slightly more efficient in the not nullable case
+    // by overloading the call operator of bin_finder to accept a value in
+    // addition to a pair and using a raw (non-pair) iterator in the
+    // transform call in detail::bin, if we need to make this faster.
+    if (input.nullable()) {
+      // Using a switch statement might be more appropriate for an enum, but it's far more verbose
+      // in this case.
+      if ((left_inclusive == inclusive::YES) && (right_inclusive == inclusive::YES))
+        return detail::bin<T, thrust::less_equal<T>, thrust::less_equal<T>, true>(
+          input, left_edges, right_edges, mr);
+      if ((left_inclusive == inclusive::YES) && (right_inclusive == inclusive::NO))
+        return detail::bin<T, thrust::less_equal<T>, thrust::less<T>, true>(
+          input, left_edges, right_edges, mr);
+      if ((left_inclusive == inclusive::NO) && (right_inclusive == inclusive::YES))
+        return detail::bin<T, thrust::less<T>, thrust::less_equal<T>, true>(
+          input, left_edges, right_edges, mr);
+      if ((left_inclusive == inclusive::NO) && (right_inclusive == inclusive::NO))
+        return detail::bin<T, thrust::less<T>, thrust::less<T>, true>(
+          input, left_edges, right_edges, mr);
+    } else {
+      if ((left_inclusive == inclusive::YES) && (right_inclusive == inclusive::YES))
+        return detail::bin<T, thrust::less_equal<T>, thrust::less_equal<T>, false>(
+          input, left_edges, right_edges, mr);
+      if ((left_inclusive == inclusive::YES) && (right_inclusive == inclusive::NO))
+        return detail::bin<T, thrust::less_equal<T>, thrust::less<T>, false>(
+          input, left_edges, right_edges, mr);
+      if ((left_inclusive == inclusive::NO) && (right_inclusive == inclusive::YES))
+        return detail::bin<T, thrust::less<T>, thrust::less_equal<T>, false>(
+          input, left_edges, right_edges, mr);
+      if ((left_inclusive == inclusive::NO) && (right_inclusive == inclusive::NO))
+        return detail::bin<T, thrust::less<T>, thrust::less<T>, false>(
+          input, left_edges, right_edges, mr);
+    }
+
+    CUDF_FAIL("Undefined inclusive setting.");
+  }
+};
+
+/// Bin the input by the edges in left_edges and right_edges.
+std::unique_ptr<column> bin(column_view const& input,
+                            column_view const& left_edges,
+                            inclusive left_inclusive,
+                            column_view const& right_edges,
+                            inclusive right_inclusive,
+                            rmm::mr::device_memory_resource* mr)
+{
+  CUDF_FUNC_RANGE()
+  CUDF_EXPECTS((input.type() == left_edges.type()) && (input.type() == right_edges.type()),
```

Comment:
The public function should call a `detail` function that has a stream parameter, then everything internal should use that stream.